src/wh_client_crypto.c

```diff
@@ -2072,4 +2072,492 @@ int wh_Client_CmacCancelableResponse(whClientContext* c, Cmac* cmac,
+                   ret);
+#endif
+            if (ret == 0) {
+                /* Response Message */
+                wh_Packet_pq_mldsa_kg_res* res     = &packet->pqMldsaKgRes;
+                uint8_t*                   key_der = (uint8_t*)(res + 1);
+                uint16_t                   res_len;
+
+                do {
+                    ret = wh_Client_RecvResponse(ctx, &group, &action, &res_len,
+                                                 (uint8_t*)packet);
+                } while (ret == WH_ERROR_NOTREADY);
+
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+                printf("[client] %s Res recv:keyid:%u, len:%u, rc:%d, ret:%d\n",
+                       __func__, res->keyId, res->len, packet->rc, ret);
+#endif
+
+                if (ret == WH_ERROR_OK) {
+                    if (packet->rc == WH_ERROR_OK) {
+                        /* Key is cached on server or is ephemeral */
+                        key_id = (whKeyId)(res->keyId);
+
+                        /* Update output variable if requested */
+                        if (inout_key_id != NULL) {
+                            *inout_key_id = key_id;
+                        }
+
+                        /* Update the context if provided */
+                        if (key != NULL) {
+                            uint16_t der_size = (uint16_t)(res->len);
+                            /* Set the key_id. Should be ERASED if EPHEMERAL */
+                            wh_Client_MlDsaSetKeyId(key, key_id);
+
+                            if (flags & WH_NVM_FLAGS_EPHEMERAL) {
+                                /* Response has the exported key */
+                                ret = wh_Crypto_MlDsaDeserializeKeyDer(
+                                    key_der, der_size, key);
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+                                wh_Utils_Hexdump(
+                                    "[client] ML-DSA KeyGen export:", key_der,
+                                    der_size);
+#endif
+                            }
+                        }
+                    }
+                    else {
+                        /* Server detected a problem with generation */
+                        ret = packet->rc;
+                    }
+                }
+            }
+        }
+        else {
+            ret = WH_ERROR_BADARGS;
+        }
+    }
+    return ret;
+}
+
+int wh_Client_MlDsaMakeCacheKey(whClientContext* ctx, int size, int level,
+                                whKeyId* inout_key_id, whNvmFlags flags,
+                                uint16_t label_len, uint8_t* label)
+{
+    if (inout_key_id == NULL) {
+        return WH_ERROR_BADARGS;
+    }
+
+    return _MlDsaMakeKey(ctx, size, level, inout_key_id, flags, label_len,
+                         label, NULL);
+}
+
+int wh_Client_MlDsaMakeExportKey(whClientContext* ctx, int level, MlDsaKey* key,
+                                 int size, WC_RNG* rng)
```

Comment:
RNG should be removed.  The server will use its own rng.

Reply:
fixed in #86